Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -0,0 +1,423 @@
+/-
+Copyright (c) 2024 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan
+-/
+import Mathlib.Algebra.Parity
+import Mathlib.Combinatorics.Enumerative.Catalan
+import Mathlib.Data.List.Indexes
+
+/-!
+# Dyck words
+
+A Dyck word is a sequence consisting of an equal number `n` of symbols of two types such that
+for all prefixes one symbol occurs at least as many times as the other.
+If the symbols are `(` and `)` the latter restriction is equivalent to balanced brackets;
+if they are `U = (1, 1)` and `D = (1, -1)` the sequence is a lattice path from `(0, 0)` to `(0, 2n)`
+and the restriction requires the path to never go below the x-axis.
+
+This file defines Dyck words and constructs their bijection with rooted binary trees,
+one consequence being that the number of Dyck words with length `2n` is `catalan n`.
+
+## Main definitions
+
+* `DyckPath`: a list of `U` and `D` steps without any restrictions.
+* `IsBalanced`: the property defining Dyck words.
+
+## Main results
+
+* `treeEquiv`: equivalence between Dyck words and rooted binary trees.
+* `finiteTreeEquiv`: equivalence between Dyck words of length `2 * n` and rooted binary trees
+  with `n` internal nodes.
+* `dyckPath_isBalanced_card_eq_catalan`: there are `catalan n` Dyck words of length `2 * n`.
```

Comment:
Do pay attention to fully qualify Lean declarations in docs. Else they won't turn into links.
```suggestion
* `DyckPath.treeEquiv`: equivalence between Dyck words and rooted binary trees.
* `DyckPath.finiteTreeEquiv`: equivalence between Dyck words of length `2 * n` and rooted binary trees
  with `n` internal nodes.
* `DyckPath.card_isBalanced_eq_catalan `: there are `catalan n` Dyck words of length `2 * n`.
```